File: opcua-webservice-spring/tutorials/lesson-2/src/main/java/org/opcfoundation/uawebservicedemo/models/company/CompanyRepositoryType.java
package org.opcfoundation.uawebservicedemo.models.company;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.uawebservicedemo.database.EmployeeDatabase;
import org.opcfoundation.uawebservicedemo.database.entity.Department;
import org.opcfoundation.uawebservicedemo.database.mapper.DepartmentMapper;
import org.opcfoundation.uawebservicedemo.models.EmployeeTwinSpace;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinRepositoryType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListResponse;
import org.opcfoundation.webserver.types.digitaltwin.DigitalTwinDescriptor;

import java.util.List;
import java.util.concurrent.CompletableFuture;

// Defining entry of company resource
public class CompanyRepositoryType extends DigitalTwinRepositoryType {
    public CompanyRepositoryType(EmployeeTwinSpace twinSpace) {
        super("CompanyRepositoryType", new LocalizedText("CompanyRepository"), twinSpace);
    }

    @Override
    public CompletableFuture<GetDigitalTwinListResponse> onGetDigitalTwinList(GetDigitalTwinListRequest request)
    {
        try
        {
            // Database query
            DepartmentMapper departmentMapper = EmployeeDatabase.getDepartmentMapper();
            List<Department> departments = departmentMapper.getChildDepartment(null); // The department without parent id is a company

            // Return department general information
            GetDigitalTwinListResponse response = new GetDigitalTwinListResponse();
            CompanyDigitalTwinType companyDigitalTwinType = ((EmployeeTwinSpace)nodeManager).companyDigitalTwinType;

            for (Department item : departments)
            {
                DigitalTwinDescriptor descriptor = new DigitalTwinDescriptor(
                        item.getId().toString(), // CompanyId
                        new LocalizedText(item.getName()), // CompanyName
                        companyDigitalTwinType); // CompanyDigitalTwinType

                response.add(descriptor);
            }

            return CompletableFuture.completedFuture(response);

        } catch (Exception e) {
            throw new UaRuntimeException(StatusCodes.Bad_NodeIdUnknown);
        }
    }
}

File: opcua-webservice-node/examples/demoserver/src/models/company/CompanyRepositoryType.ts
import { StatusCodes } from "opcua-webapi";
import { UaError, UaLocalizedText, makeUaStatusCode } from "opcua-webapi-ts";
import {
    DigitalTwinDescriptor,
    DigitalTwinRepositoryType,
    GetDigitalTwinListRequest,
    GetDigitalTwinListResponse,
} from "opcua-webservice-node";

import { prisma } from "../../connectors/prismaClient";
import { EmployeeTwinSpace } from "../EmployeeTwinSpace";

export class CompanyRepositoryType extends DigitalTwinRepositoryType {
    constructor(twinSpace: EmployeeTwinSpace) {
        super("CompanyRepositoryType", new UaLocalizedText("CompanyRepository"), twinSpace);
    }

    public async onGetDigitalTwinList(request: GetDigitalTwinListRequest): Promise<GetDigitalTwinListResponse> {
        // Database query
        const departments = await prisma.department.findMany({
            where: { ParentDepartmentId: null } // The department without parent id is a company
        });

        // Return department general information
        const response = new GetDigitalTwinListResponse();
        const companyDigitalTwinType = EmployeeTwinSpace.CompanyDigitalTwinType;

        for (const item of departments) {
            const descriptor = new DigitalTwinDescriptor(
                item.ID.toString(), // CompanyId
                new UaLocalizedText(item.Name), // CompanyName
                companyDigitalTwinType // CompanyDigitalTwinType
            );

            response.add(descriptor);
        }

        return response;
    }
}